### DISCUSSION TITLE
how to increase max_request_size at request level

### DISCUSSION THREAD

**Discussion by rmallah** [mojolicious/mojo#1865] [Q&A]
We want to increase max_request_size only for the requests that 
handle the file upload in a certain controller.

```
sub upload {
    my $c = shift ;

  # how do i increase max_request_size here.
   ......
  ......
}
```

https://docs.mojolicious.org/Mojolicious#max_request_size

above describes how to do in startup {} , but we do not want 
to increase the limit for the entire app.

regds
Rajesh Mallah.

**Comment by Grinnz** [comment]
It's possible you'd be able to increase instead `$c->req->max_message_size` in the action, but it would have to occur before anything attempts to parse the request body or it may already have been applied.

**Reply by rmallah** [reply]
it does not help.

also making the below message more verbose to indicate the current threshold
would be more informative. Shouldnt' it be a 'warn' instead of 'debug' ?
should we open an issue for this ?

```
[2021-10-25 14:11:21.71773] [1774] [debug] Maximum message size exceeded                                                                             
```

**Reply by kraih** [reply]
It has to be done earlier. From an event that triggers before the message body is received. https://docs.mojolicious.org/Mojolicious/Guides/Cookbook#Streaming-multipart-uploads

**Reply by rmallah** [reply]
the below works (i have tested the code for positive and negative cases)

```

sub startup {
 my $app =  shift;

 ........

 $app->hook(after_build_tx => sub  {
    my $tx =  shift;
    $tx->on(request => sub  {
      my $url = shift->req->url ;
      if ($url =~ /upload/) {
        $app->max_request_size(104857600);
      }   
    }); 
  });

  ........
}

```
Q: how do i restore the limit after the request is done ?

**Reply by Grinnz** [reply]
You can set it only for that request by setting max_message_size on $c->req instead of on the application. Requests may not happen sequentially so if you set it on the app it may affect other requests no matter how soon you change it back.

**Reply by rmallah** [reply]
I checked below does _not_ work.

```
 $app->hook(after_build_tx => sub  {
    my $tx =  shift;
    my $app = shift;
    $tx->on(request => sub  {
      my $req = shift->req ;
      if ($req->url =~ /upload/) {
        $req-> max_message_size (104857600);
      }   
    }); 
  }); 

```

**Reply by rmallah** [reply]
> You can set it only for that request by setting max_message_size on $c->req instead of on the application. Requests may not happen sequentially so if you set it on the app it may affect other requests no matter how soon you change it back.

I am not sure if a strict control is possible, but an eventually consistent threshold converging to default value is still better 
and having it (app level setting) raised till lifetime of the process.
